Embed Socials feature and other improvements

WombatFromHell opened the pull request:
The main things in this PR are:

- Minor styling changes to the Who-LOL'd button (to better fit with the post box)
- More performance hacks for Webkit
- Giphy, Gfycat, and Imgur have better parsing and embedding
- Who-LOL'd buttons now toggle themselves when re-clicked and show multiple LOL'd categories at once (like ThomW's script)
- (Updated: Improvements to how Embed Images works when embedding and positioning embedded elements in the postbox. Should use screen real-estate more efficiently now.)

Now, as for the "Embed Socials" script, I need to go over a few things in this comment:

1) Twitter support works as advertised - nothing to write about there
2) Instagram support is limited to whatever links Instagram's wjs script will properly render (due to CORS issues). Unless I'm missing something, this seems pretty hit or miss in general when letting wjs embed `<blockquote>` content. I've yet to find a better way that preserves formatting when it actually -does- work, but I'm always open to suggestions if anyone else thinks they can improve the implementation.

WombatFromHell commented:
Ahh, turns out the `X-Frame-Options` CORS issue is actually due to server checking if the profile is public or private and setting `X-Frame-Options: SAMEORIGIN` if its private. Given that I lack a back-end to use as a proxy I don't think there's a reliable way I can do a preflight check on the endpoint to see if `X-Frame-Options: SAMEORIGIN` comes back from a given Instagram link before inlining it.

I'll think about it for a bit and respond back if I find another solution.

electroly commented on chromeshack.css:
```diff
@@ -788,6 +869,54 @@ a.refresh_pending {
     height: 15px !important;
 }
 
+.postmeta a[id^="get_lol_users_"] {
```
Any concern on browser performance with selectors like these?  If it's our own elements, why not stick a class on it?

WombatFromHell commented:
The toggling logic that prevents reprocessing a link is tied to those selectors. I haven't noticed any major performance hit, but I haven't comprehensively tested for it.

WombatFromHell commented:
I'm also seeing some styling weirdness with the Pending Posts button causing the buttons on the nav bar at the top-right to shift when it unhides.

WombatFromHell commented:
I'm also tracking down some bugs with Imgur parsing in `image_loader.js`.

WombatFromHell commented:
Okay, I think that's enough tweaking.

electroly commented:
What's the issue with the Instagram private profiles?  Does that mean if someone posts a link to a private Instagram profile, or does it mean if the Chrome Shack user is logged into a private profile?

WombatFromHell commented:
If someone posts a link to a post from a private Instagram profile the Instagram endpoint will deny the request to render the post content (via X-Frame-Options). AFAIK, there is no reliable way (without a backend) to circumvent that.

A good example is Galadriel's Instagram posts on the Chatty. She has a private profile so when a link is posted it will refuse to show the media and direct you to the main website. Other public Instagram profiles will work fine though.

electroly commented:
Galadriel is almost the only person who posts Instagram links; is it worth having the feature if it won't work on those links?

WombatFromHell commented:
Like I said, it's an Instagram server issue when using their wjs inlining method. Inlining works fine for everything else. I'll keep researching alternative rendering methods to see if I can work around the Instagram server's limitations when using their preferred embedding method.

WombatFromHell commented:
There are workarounds, but they involve implementing standalone rendering logic and relying on the JSON endpoints or scraping the Instagram page directly.

There are projects out there that render from the endpoints so I'll look into how their implementations are done and see how I can improve over using the widget.

WombatFromHell commented:
I'm nearly ready to push the updated Instagram embedder, but I'm stuck chasing down bugs with the video styling not being properly responsive. I should have it done by this evening.

WombatFromHell commented:
These two threads are good ways to check the new Instagram embedder (for reference):

- https://www.shacknews.com/chatty?id=38076395#item_38076395

- https://www.shacknews.com/chatty?id=37947143#item_37947143

I think the embedding logic should handle multi-image/video posts okay (it will just default to the first piece of media in the set), but I wasn't able to find any examples on the Chatty of any that don't link straight to a gallery.

electroly commented on options.html:
```diff
@@ -99,6 +99,11 @@ <h2>
+                    <input type="checkbox" class="script_check" id="embed_socials" />
+                    <label for="video_loader">Embed Social Links</label>
+                </p>
+                <p class="indent info">Replace links to Twitter and Instagram with embedded posts (note: disable tracker blocking).</p>
```
Can you expand the options text to explain what "disable tracker blocking" means?  If it's required, it would be nice if it had a link to some instructions on how to do it in Chrome and Firefox.  I wasn't sure what this was asking me to do.

WombatFromHell replied:
Sure. Basically, Firefox optionally lets you prevent javascript and cookie trackers from running, and that conflicts with Twitter/Instagram embedding. I don't think Chrome does this at the moment unless you're using an extension like Ghostery or some such.

I'll add a more detailed blurb about it in the feature option on the options page.

electroly commented on options.html:
```diff
@@ -99,6 +99,11 @@ <h2>
                 </div>
+                <p>
+                    <input type="checkbox" class="script_check" id="embed_socials" />
+                    <label for="video_loader">Embed Social Links</label>
```
Wrong label target here; if you click the "Embed Social Links" text it checks "Embed videos" instead.

WombatFromHell replied:
Ahhh, good catch.

WombatFromHell commented:
I agree with your point about letting the user know what a link will do before they click on it. I was thinking along the lines of a little icon on the side of the link that indicates it will be embedded inline. Does this sound reasonable?

After I changed how the image/video embedder styles embeds I realized that it makes sense to have some sort of overlay to re-collapse them once they've been embedded (in addition to seeing what's been opened via highlighting, which is already functional). Before, clicking the image/video would hide it, but I refactored a lot of that so I should probably include that functionality again to at least bring it back to feature parity with the previous code.

electroly commented:
Yeah an icon sounds good.  I wonder if a simple [+] and [-] graphic to indicate expand/collapse would do the trick.

WombatFromHell commented:
I've got "Embed Images" working with an expando button next to every "hooked" link. It may take me another day to get "Embed Videos" and "Embed Socials" working in the same way. Example:

https://i.imgur.com/9bUG85r.gifv

Will this work?

electroly commented:
Looks great!  I like that a lot.

wjrogers commented:
That's really nice, takes the guesswork out of remembering which feature
you have enabled.

On Sun, Nov 18, 2018, 11:08 AM Brian Luft <notifications@github.com> wrote:

> Looks great! I like that a lot.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/arhughes/chromeshack/pull/98#issuecomment-439704067>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAD0bfMnhv5nvX0nC8ogCEQ7WcXI_mt1ks5uwYYBgaJpZM4YdE97>
> .
>

electroly commented:
What do you think about combining the options for all the embed scripts into a single "embed media links" option?  I wonder if people really want to enable some of the embed scripts but not others.

Additionally I'm wondering if we should enable this by default.  It's pretty unobtrusive.

wjrogers commented:
I only enable embed images. I don't like embedding video.

The Twitter embed has always annoyed the shit out of me, but only because
it never ever worked. If it's fixed then I'd probably use that, too.

On Sun, Nov 18, 2018, 11:16 AM Brian Luft <notifications@github.com> wrote:

> What do you think about combining the options for all the embed scripts
> into a single "embed media links" option? I wonder if people really want to
> enable some of the embed scripts but not others.
>
> Additionally I'm wondering if we should enable this by default. It's
> pretty unobtrusive.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/arhughes/chromeshack/pull/98#issuecomment-439704612>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAD0bQp20WcwRs-1wXokyHni6eYI9TJeks5uwYfPgaJpZM4YdE97>
> .
>

electroly commented:
Makes sense.  I've never used any of these embed scripts so I have no real opinion.  I wonder if we could get the best of both worlds by only expanding the inline embed if you click the expando button, and the regular link works the same as before.  Since I never used embedding this seems great to me but it may be too annoying for embed users who are used to clicking the link itself.

wjrogers commented:
Small target vs big target. I know we're all 360 noscope headshotters
around here but sometimes keeping it simple is better. You can still
right-click copy link or open in new tab.

On Sun, Nov 18, 2018, 12:02 PM Brian Luft <notifications@github.com> wrote:

> Makes sense. I've never used any of these embed scripts so I have no real
> opinion. I wonder if we could get the best of both worlds by only expanding
> the inline embed if you click the expando button, and the regular link
> works the same as before. Since I never used embedding this seems great to
> me but it may be too annoying for embed users who are used to clicking the
> link itself.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/arhughes/chromeshack/pull/98#issuecomment-439708109>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAD0bSvtI3sb8Se2hfWO7Z5Im6DurDI7ks5uwZKkgaJpZM4YdE97>
> .
>

WombatFromHell commented:
I like providing users with as many options as possible as long as it doesn't clutter the page. Having every embed stack vertically and ruin the flow of the text was annoying, that's why I have it set this way.

As for Twitter embeds, my preference would be to have a simplified box with the content of the tweet rather than a wjs embedded iframe. That would require lots of parsing, API calls, and work that I don't have time for, honestly.

Anyway, I had to refactor most of the video embedding code too. I think the improved implementation is more flexible (channels, videos, clips, and playlists work now).

I should be done as soon as I'm finished porting the Embed Socials code to the new media container method. So, probably this afternoon, depending on how many bugs I find going back over the social embedding code.

WombatFromHell commented:
Also, I would rather take the time to tweak the CSS to get the fitting of content to work more cleanly than to have two options for positioning.

The way it's currently set the CSS grid will start with a single column that spans the whole post box, and split into two columns if the content allows. I need to tweak it a bit more to get it closer to the behavior of the old styling, but with the improvements of the new method. Shouldn't be too difficult.

WombatFromHell commented:
I've noticed some bugs in Youtube embedding. I'm tracking it down now.

WombatFromHell commented:
In theory this last few commits should fix the remaining bugs that I could spot in state transitions, embedding/parsing, and container styling. Let me know if I missed anything.

WombatFromHell commented:
I'm having some difficulty with styling the media container. I to fit 3 columns max into the width of the container while still limiting max-height and max-width per item to prevent stretching embeds too much. Currently, I'm using `display: grid` and `auto-fit` columns/rows which gets me 95% of the way there, but it has this irritating side effect when combined with `object-fit: contain` on image/video embeds: it places empty space on the side of each child item in the container.

The other alternative that works, but also has problems is using `display: flex` with `flex-wrap: wrap` and then putting `flex: 0 1 auto` on the children. The downside is that I can't easily affix a small gutter to each item while preserving responsiveness, unless I add another element 'wrapper' to each child.

I was wondering if anyone could give advice on this problem. It's pretty much the last thing blocking me from signing off on this fork.

electroly commented:
> unless I add another element 'wrapper' to each child.

Is the only problem with this alternative that it isn't a clean solution, or is there another problem?  If you can hack up the HTML to make it work, that's what I'd do.

WombatFromHell commented:
The flex method has some shortcomings besides needing element wrappers, but they're not as obnoxious as the limitations of the grid method. A hammer for a nail I guess.

I guess I shouldn't take shortcuts and just take the time to go through my HTML see if I can improve how everything fits since I've already migrated all the embeds over to the `media-container` method. The way I was doing it before was intended to not interfere with embeds, but that isn't a problem anymore I just have to be careful with `iframe` elements like Youtube, Twitch, and Twitter.

WombatFromHell commented:
Okay, so, I've got two things blocking progress on finishing this PR, and one other thing I need to mention:

1) I'm struggling with a race condition that occurs when double clicking handled media links, due to the async nature of how network requests and rendering are currently done. I need some more time to polish it up.

2) I still can't get a layout that works for image/video embeds of widely varying aspect ratios without odd problems. Neither grid nor flexbox will work with all cases by themselves, it seems additional javascript and resize handlers are necessary to update the items in the media container. Essentially, what I'm after is something similar to the Pinterest layout (also known as the Masonry layout). I'll need to take some more time to hammer out something that will work cleanly with the vast majority of aspect-ratio cases.

3) Now for the note: in my rollup (commit d5a7db0) I've started breaking styles out into individual files, roughly by script/feature, to help in future code maintenance. I was getting annoyed with scrubbing through one huge CSS file so I figured it made sense to break that stuff out. I've been thinking about putting media queries into yet another file so one doesn't have to search the code-base for them (they currently hang out in `styles/media.css`).

All that aside, commit d031f15 is pretty close to late beta code (all features are working but some, like the race condition bug, can have issues sometimes). These polishing changes may take me another week or two.

WombatFromHell commented:
Okay, this commit is another rollup with many fixes and improvements. After a lot of testing, experimenting, and trial-and-error I've come to the conclusion that using a browser addon to inject and manage the code necessary for a Masonry style layout is too much of a hassle. So, instead, I think I've worked out a grid layout that should work, though it isn't as pretty as using masonry.

Basically, it's three layouts based on screen width and leaning on grid's inherent responsiveness. A column-only layout for narrow screenspace, a video-first-or-50% layout for most non-fullscreen situations, and a larger-or-33% layout for fullscreen.

Hopefully this commit is a bit more descriptive about what's been changed than the last rollup.

WombatFromHell commented:
Okay, I -think- I've got the Imgur API nonsense and the styling stuff worked out. After spending a considerable amount of time debugging my way around it, I can't believe how crappy Imgur's API is.

Anyway, I'll be sticking to bug testing unless an issue is reported.

WombatFromHell commented:
I've been hammering on the current build of this PR over the weekend and I'm not seeing any obvious bugs (after I got the Gfycat stuff sorted that is). I think I'm pretty much done unless someone wants to offer criticisms of other things I need to work on here.

Anyway, I'll move on to working on issue #97. I've already got Custom User Filters moved over to using the new `MutationObserver()` events (leaning on a new `processRootPostEvent` in `chromeshack_posts.js`) rather than crawling the DOM, next I'll port the other userId dependent scripts over to using the `getUserId()` helper function. Should be pretty straight forward.

electroly commented:
The issue with the label in the options page appears to still be there.  The functionality looks good to me.  I like the toggle button.  When do you want to merge this?

WombatFromHell commented:
Err... If you're talking about the Embed Socials option description, I recall changing the Embed Socials option label around the time I did my second commit rollup. It's currently:

"Replace links to Twitter and Instagram with embedded posts.

NOTE: Tracker blocking addons like Ghostery and Firefox's Tracker Blocker interfere with this feature. Please disable tracker blocking!"

Would you suggest something more specific or am I misunderstanding?

electroly commented:
I mean when I click on the "Embed Social Links" text, it toggles the "Embed videos" checkbox instead of the one for social links due to the wrong label target.

WombatFromHell commented:
Oops, that must have been a regression caused by a local merge in one of my rollups a few weeks back. I swear I did fix that at one point.

electroly commented:
The first two things on the first page of the chatty that I clicked after enabling this to test were:

- A link to an Instagram picture of fuzzy cat balls.
- A link to an Imgur picture of those cat balls photoshopped into a Sonic the Hedgehog movie poster.

I don't know what I expected.

WombatFromHell commented:
LOL, now that's a Christmas present that keeps on giving.

Seriously though, there are obviously going to be some bugs in the Instagram container, despite my best efforts. I'll investigate and see if I can do a smaller and more targeted PR with bug fixes specifically limited to embed behavior. I think what I'll do is throw this into an issue ticket so it can be more easily documented in one place.

Also, I feel like I've been getting a bit carried away with straying outside the bounds of the feature(s) I'm working on, which is probably making testing/verification of PRs more messy than it has to be. I'll try to split things into discrete PRs and make commits more transparent from now on.

electroly commented:
We can merge this, it's fine.  I don't see any real problems.  People can disable the script if anything weird happens.
